File: frontend/src/api/client/lexware-documents.ts
import { API_BASE, ApiError, getAuthToken, parseContentDispositionFilename, request } from './core';

export interface LexwarePaymentItem {
  item_type: string;
  category: string;
  amount: string;
  currency: string;
  posting_date: string | null;
}

export interface LexwareDocumentFinance {
  currency: string;
  total_amount: string | null;
  open_amount: string | null;
  payment_state: 'known' | 'unknown' | 'not_applicable';
  payment_status: string | null;
  direction: 'receivable' | 'payable' | 'none';
  credit: boolean;
  overdue: boolean | null;
  included_in_totals: boolean;
  exclusion_reason: string | null;
  payment_items: LexwarePaymentItem[];
}

export interface LexwareOriginalFile {
  file_id: string;
  cached: boolean;
  filename?: string | null;
  media_type?: string | null;
  size_bytes?: number | null;
  sha256?: string | null;
  cached_at?: string | null;
}

export interface LexwareDocument {
  id: number;
  connection_id: number;
  business_profile_id: number;
  company_name: string;
  source: 'lexware';
  external_id: string;
  voucher_type: string;
  voucher_status: string;
  voucher_number: string | null;
  voucher_date: string | null;
  contact_name: string | null;
  supported: boolean;
  archived: boolean;
  in_latest_sync: boolean;
  connection_enabled: boolean;
  sync_status: string;
  last_success_at: string | null;
  updated_at: string;
  version: number;
  local_document_id: number | null;
  due_date?: string | null;
  finance?: LexwareDocumentFinance;
  files?: LexwareOriginalFile[];
}

export interface LexwareFinance {
  source: 'lexware';
  as_of: string;
  totals: Array<{
    currency: string;
    receivables: string;
    payables: string;
    overdue_receivables: string;
    overdue_payables: string;
    document_count: number;
  }>;
  included_count: number;
  linked_count: number;
  unknown_count: number;
  excluded_count: number;
  unsupported_count: number;
  stale_connection_count: number;
}

export interface LexwareDocumentFilters {
  connection_id?: number;
  business_profile_id?: number;
  search?: string;
  voucher_type?: string;
  offset?: number;
  limit?: number;
}

function query(filters: LexwareDocumentFilters): string {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(filters)) {
    if (value !== undefined && value !== '') params.set(key, String(value));
  }
  return params.size ? `?${params}` : '';
}

export const lexwareDocumentsApi = {
  list: (filters: LexwareDocumentFilters = {}) =>
    request<{ items: LexwareDocument[]; total: number }>(`/lexware/documents${query(filters)}`),
  detail: (id: number) => request<LexwareDocument>(`/lexware/documents/${id}`),
  finance: (filters: Pick<LexwareDocumentFilters, 'connection_id' | 'business_profile_id'> = {}) =>
    request<LexwareFinance>(`/lexware/finance${query(filters)}`),
  link: (id: number, localDocumentId: number | null, expectedVersion: number) =>
    request<LexwareDocument>(`/lexware/documents/${id}/link`, {
      method: 'PUT', body: JSON.stringify({ local_document_id: localDocumentId, expected_version: expectedVersion }),
    }),
  download: async (documentId: number, fileId: string) => {
    const token = getAuthToken();
    const response = await fetch(`${API_BASE}/lexware/documents/${documentId}/files/${encodeURIComponent(fileId)}`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!response.ok) {
      const body = await response.json().catch(() => ({}));
      throw new ApiError(typeof body.detail === 'string' ? body.detail : `HTTP ${response.status}`, response.status);
    }
    return {
      blob: await response.blob(),
      filename: parseContentDispositionFilename(response.headers.get('Content-Disposition')) || `lexware-${fileId}`,
    };
  },
};

File: backend/app/schemas/lexware_documents.py
"""Explicit public projections: raw upstream payloads are never returned."""

from datetime import date
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from backend.app.schemas.utc_timestamp import UtcTimestamp


class LexwarePaymentItemRead(BaseModel):
    item_type: str
    category: str
    amount: Decimal
    currency: str
    posting_date: date | None


class LexwareDocumentFinanceRead(BaseModel):
    currency: str
    total_amount: Decimal | None
    open_amount: Decimal | None
    payment_state: Literal["known", "unknown", "not_applicable"]
    payment_status: str | None
    direction: Literal["receivable", "payable", "none"]
    credit: bool
    overdue: bool | None
    included_in_totals: bool
    exclusion_reason: str | None
    payment_items: list[LexwarePaymentItemRead]


class LexwareFileRead(BaseModel):
    file_id: str
    cached: bool
    filename: str | None = None
    media_type: str | None = None
    size_bytes: int | None = None
    sha256: str | None = None
    cached_at: UtcTimestamp | None = None


class LexwareDocumentRead(BaseModel):
    id: int
    connection_id: int
    business_profile_id: int
    company_name: str
    source: Literal["lexware"] = "lexware"
    external_id: str
    voucher_type: str
    voucher_status: str
    voucher_number: str | None
    voucher_date: date | None
    contact_name: str | None
    supported: bool
    archived: bool
    in_latest_sync: bool
    connection_enabled: bool
    sync_status: str
    last_success_at: UtcTimestamp | None
    updated_at: UtcTimestamp
    version: int
    local_document_id: int | None
    # These keys are omitted entirely without payments:read.
    due_date: date | None = None
    finance: LexwareDocumentFinanceRead | None = None
    files: list[LexwareFileRead] | None = None


class LexwareDocumentList(BaseModel):
    items: list[LexwareDocumentRead]
    total: int


class LexwareLinkCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")
    local_document_id: int | None = Field(ge=1)
    expected_version: int = Field(ge=1)


class LexwareCurrencyTotals(BaseModel):
    currency: str
    receivables: Decimal = Decimal("0.00")
    payables: Decimal = Decimal("0.00")
    overdue_receivables: Decimal = Decimal("0.00")
    overdue_payables: Decimal = Decimal("0.00")
    document_count: int = 0


class LexwareFinanceRead(BaseModel):
    source: Literal["lexware"] = "lexware"
    as_of: date
    totals: list[LexwareCurrencyTotals]
    included_count: int
    linked_count: int
    unknown_count: int
    excluded_count: int
    unsupported_count: int
    stale_connection_count: int
